packages/helix-shared-bounce/src/bounce.js

```diff
@@ -29,17 +37,41 @@ function bounce(func, { responder, timeout = 500 }) {
     context.invocation.bounceId = bounceId;
 
     // run the quick responder function
-    const holdingResponse = await responder(request, context);
+    const holdingResponse = (async () => {
+      const res = responder(request, context);
+      const wait = timer.setTimeout(timeout);
+      try {
+        const response = await Promise.all([res, wait]);
+        return response[0];
+      } catch (err) {
+        return new Response('Internal Server Error', { status: 500 });
+      }
+    })();
     // invoke the current function again, via HTTP, with the x-hlx-bounce-id
     // header set, so that we don't get into an endless loop
     request.headers.set('x-hlx-bounce-id', bounceId);
     const signal = timeoutSignal(2 * timeout);
-    const actualResponse = fetch(request, {
-      signal,
-    });
+    const actualResponse = (async () => {
+      try {
+        return await fetch(request, {
+          signal,
+        });
+      } catch (e) {
+        if (e instanceof AbortError) {
+          return new Response(e.message, {
+            // we acted as a gateway, but the upstream was too slow
+            status: 504,
+          });
+        }
+        context.log.warn(`error while bouncing: ${e.message}`);
+        return new Response(e.message, {
+          // we acted as a gateway, but there was an error with the network connection
+          status: 502,
+        });
+      }
```

Comment:
you need to add a `finally { signal.clear() } ` otherwise the process might hang.

Reply:
done as well.